go/libkb/api.go

```diff
@@ -262,6 +266,76 @@ func doRequestShared(api Requester, arg APIArg, req *http.Request, wantJSONRes b
 	return internalResp, jw, nil
 }
 
+// doRetry will just call cli.cli.Do if arg.Timeout and arg.RetryCount aren't set.
+// If they are set, it will cancel requests that last longer than arg.Timeout and
+// retry them arg.RetryCount times.
+func doRetry(arg APIArg, cli *Client, req *http.Request) (*http.Response, error) {
+	if arg.Timeout == 0 && arg.RetryCount == 0 {
+		return cli.cli.Do(req)
+	}
+
+	timeout := cli.cli.Timeout
+	if arg.Timeout != 0 {
+		timeout = arg.Timeout
+	}
+
+	retries := 1
+	if arg.RetryCount > 1 {
+		retries = arg.RetryCount
+	}
+
+	var lastErr error
+	for i := 0; i < retries; i++ {
+		if i > 0 {
+			arg.G().Log.Debug("retry attempt %d of %d for %s", i, retries, arg.Endpoint)
+		}
+		resp, err := doTimeout(cli, req, timeout)
+		if err == nil {
+			return resp, nil
+		}
+		lastErr = err
+	}
+
+	return nil, fmt.Errorf("doRetry failed, attempts: %d, timeout %s, last err: %s", retries, timeout, lastErr)
+
+}
+
+// adapted from https://blog.golang.org/context httpDo func
+func doTimeout(cli *Client, req *http.Request, timeout time.Duration) (*http.Response, error) {
+	// TODO: could pass in a context from further up the chain
+	// and use that instead of context.Background()
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	// new way to cancel requests
+	reqCancel := make(chan struct{})
+	req.Cancel = reqCancel
+
+	type response struct {
+		resp *http.Response
+		err  error
+	}
+	c := make(chan response, 1)
+	go func() {
+		var r response
+		r.resp, r.err = cli.cli.Do(req)
+		c <- r
+	}()
+
+	select {
+	case <-ctx.Done():
```

Comment:
yeah, basically.

also, go 1.7 net/http has context built in, so this will prepare us for that as well.